compiler/natives/src/reflect/reflect.go

```diff
@@ -550,21 +550,62 @@ type mapIter struct {
 	m    *js.Object
 	keys *js.Object
 	i    int
+	last *js.Object
 }
 
-func mapiterinit(t *rtype, m unsafe.Pointer) *byte {
-	return (*byte)(unsafe.Pointer(&mapIter{t, js.InternalObject(m), js.Global.Call("$keys", js.InternalObject(m)), 0}))
+func mapiterinit(t *rtype, m unsafe.Pointer) unsafe.Pointer {
+	return unsafe.Pointer(&mapIter{t, js.InternalObject(m), js.Global.Call("$keys", js.InternalObject(m)), 0, nil})
 }
 
-func mapiterkey(it *byte) unsafe.Pointer {
-	iter := (*mapIter)(unsafe.Pointer(it))
-	k := iter.keys.Index(iter.i)
-	return unsafe.Pointer(js.Global.Call("$newDataPointer", iter.m.Get(k.String()).Get("k"), jsType(PtrTo(iter.t.Key()))).Unsafe())
+func mapiterkey(it unsafe.Pointer) unsafe.Pointer {
+	iter := (*mapIter)(it)
+	var kv *js.Object
+	if iter.last != nil {
+		kv = iter.last
+	} else {
+		// Compare the index and the size of the actual key set, and check if the iterator is already exhausted.
+		if iter.i >= js.Global.Call("$keys", iter.m).Length() {
```

Comment:
Calling `$keys` again does not look right at first glance. Which bug is this supposed to fix?

Reply:
* [`reflect.TestMapIterDelete1`](https://github.com/golang/go/blob/4fb900e9ca1f08c57b074e7bf6a7eab90b92c898/src/reflect/all_test.go#L6686-L6700)

`Value` tries to access `mapiterkeys` at the second and later `Next` call ([code](https://github.com/golang/go/blob/4fb900e9ca1f08c57b074e7bf6a7eab90b92c898/src/reflect/value.go#L1197), and `iter.keys` can be already not synced with `$keys` by deleting keys in a loop.

On second thought, my fix might not work with the below code:

```go
iter := ValueOf(m).MapRange()
for iter.Next() {
    delete(m, iter.Key())
}
if len(m) != 0 {
    panic("not reached")
}
```

Let me think...

Reply:
Fixed.